plugins/gnupgplugin/gpgmebase.cpp

```diff
@@ -0,0 +1,761 @@
+    if (email == kpgParams.customParameters().constEnd()) {
+        return Result(Result::CryptoPluginKeyGenerationError,
+                      QStringLiteral("missing email custom parameter."));
+    }
+
+    GPGmeContext ctx(m_protocol, home);
+    if (!ctx)
+        return Result(Result::CryptoPluginKeyGenerationError, ctx.error());
+
+    QString gnupgKeyParms = "<GnupgKeyParms format=\"internal\">\n";
+    if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairDsa) {
+        gnupgKeyParms += "Key-Type: DSA\n";
+        gnupgKeyParms += "Key-Length: 1024\n";
+        gnupgKeyParms += "Subkey-Type: DSA\n";
+        gnupgKeyParms += "Subkey-Length: 1024\n";
+    } else if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairRsa) {
+        gnupgKeyParms += "Key-Type: RSA\n";
+        gnupgKeyParms += QStringLiteral("Key-Length: %1\n").arg(static_cast<RsaKeyPairGenerationParameters>(kpgParams).modulusLength());
+        gnupgKeyParms += "Subkey-Type: RSA\n";
+        gnupgKeyParms += QStringLiteral("Subkey-Length: %1\n").arg(static_cast<RsaKeyPairGenerationParameters>(kpgParams).modulusLength());
+    } else {
+        gnupgKeyParms += "Key-Type: default\n";
+    }
+    gnupgKeyParms += "Key-Usage: sign";
+    if (keyTemplate.operations() & CryptoManager::OperationEncrypt)
+        gnupgKeyParms += ", encrypt";
+    gnupgKeyParms += "\n";
+    QVariantMap::ConstIterator passphrase = kpgParams.customParameters().constFind("passphrase");
+    if (passphrase != kpgParams.customParameters().constEnd())
+        gnupgKeyParms += QStringLiteral("Passphrase: %1\n").arg(passphrase->toString());
+    if (m_protocol == GPGME_PROTOCOL_OpenPGP) {
+        gnupgKeyParms += QStringLiteral("Name-Real: %1\n").arg(name->toString());
+        QVariantMap::ConstIterator comment = kpgParams.customParameters().constFind("comment");
+        if (comment != kpgParams.customParameters().constEnd())
+            gnupgKeyParms += QStringLiteral("Name-Comment: %1\n").arg(comment->toString());
+    }
+    gnupgKeyParms += QStringLiteral("Name-Email: %1\n").arg(email->toString());
+    QVariantMap::ConstIterator expire = kpgParams.customParameters().constFind("expire");
+    if (expire != kpgParams.customParameters().constEnd())
+        gnupgKeyParms += QStringLiteral("Expire-Date: %1\n").arg(expire->toString());
+    gnupgKeyParms += QStringLiteral("</GnupgKeyParms>");
+
+    gpgme_error_t err;
+    if (m_protocol == GPGME_PROTOCOL_OpenPGP) {
+        err = gpgme_op_genkey(ctx, gnupgKeyParms.toUtf8().constData(), 0, 0);
+    } else {
+        GPGmeData pub, priv;
+        if (!pub || !priv) {
+            return Result(Result::CryptoPluginKeyGenerationError,
+                          QStringLiteral("cannot create data."));
+        }
+ 
+        err = gpgme_op_genkey(ctx, gnupgKeyParms.toUtf8().constData(), pub, priv);
+
+        // Todo do something with the public data.
```

Comment:
is the public data not already used above with the gpgme_op_genkey() call?

Reply:
I've updated the comment. Here it's to do something with the generated data from gpgme that is a certificate request in S/MIME.